table_stats.go

```diff
@@ -583,3 +579,162 @@ func estimateEntrySizes(
+// As an example, consider the following set of spans from the range deletion
+// and range key blocks of a table:
+//
+//         |---------|     |---------|         |-------| RANGEKEYDELs
+//   |-----------|-------------|           |-----|       RANGEDELs
+// __________________________________________________________
+//   a b c d e f g h i j k l m n o p q r s t u v w x y z
+//
+// The tableRangedDeletionIter produces the following set of output spans, where
+// '1' indicates a span containing only range deletions, '2' is a span
+// containing only range key deletions, and '3' is a span containing a mixture
+// of both range deletions and range key deletions.
+//
+//      1       3       1    3    2          1  3   2
+//   |-----|---------|-----|---|-----|     |---|-|-----|
+// __________________________________________________________
+//   a b c d e f g h i j k l m n o p q r s t u v w x y z
+//
+// Algorithm.
+//
+// The iterator first defragments the range deletion and range key blocks
+// separately. During this defragmentation, the range key block is also filtered
+// so that keys other than range key deletes are ignored. The range delete and
+// range key delete keyspaces are then merged.
+//
+// Note that the only fragmentation introduced by merging is from where a range
+// del span overlaps with a range key del span. Within the bounds of any overlap
+// there is guaranteed to be no further fragmentation, as the constituent spans
+// have already been defragmented. To the left and right of any overlap, the
+// same reasoning applies. For example,
+//
+//            |--------|         |-------| RANGEKEYDEL
+//   |---------------------------|         RANGEDEL
+//   |----1---|----3---|----1----|---2---| Merged, fragmented spans.
+// __________________________________________________________
+//   a b c d e f g h i j k l m n o p q r s t u v w x y z
+//
+// Any fragmented abutting spans produced by the merging iter will be of
+// differing types (i.e. a transition from a span with homogenous key kinds to a
+// heterogeneous span, or a transition from a span with exclusively range dels
+// to a span with exclusively range key dels). Therefore, further
+// defragmentation is not required.
+//
+// Each span returned by the tableRangeDeletionIter will have at most four keys,
+// corresponding to the largest and smallest sequence numbers encountered across
+// the range deletes and range keys deletes that comprised the merged spans.
+type tableRangedDeletionIter struct {
+	keyspan.FragmentIterator
+}
+
+// newTableRangeDeletionIter returns a new tableRangedDeletionIter.
+func newTableRangeDeletionIter(
+	cmp base.Compare, r *sstable.Reader, m *fileMetadata,
+) (*tableRangedDeletionIter, error) {
+	// The range del iter and range key iter are each wrapped in their own
+	// defragmenting iter. For each iter, abutting spans can always be merged.
+	var equal keyspan.DefragmentMethodFunc = func(_ base.Compare, a, b keyspan.Span) bool { return true }
```

Comment:
Bummer. Turns out in versions of Go prior to 1.18 these can't be inlined. Will amend.

Reply:
Actually ... that seems to be fake news. Interestingly, what I had originally compiles just fine for me locally, but it's failing in CI. Lifting the functions out didn't do anything either.

Reply:
Fixed. Dodgy rebase didn't pick up the `keyspan.Span` to `*keyspan.Span` change.